### ISSUE TITLE
Provide option in ForEach-Object -parallel to transfer current runspace state

### ISSUE THREAD

**Issue by PaulHigin** [PowerShell/PowerShell#12240]
# Summary of the new feature/enhancement

Foreach-Object -parallel currently runs each loop iteration in a runspace initialized to its default state.  But users might reasonably expect the runspace state to be the same as the runspace state running the foreach-object cmdlet.

This feature idea is to add a new `-UseCurrentState` (or something) switch that transfers the current runspace state to each parallel loop iteration runspace.

This would ensure that any modules imported or defined functions are available to each foreach -parallel script.  

It is not clear how much of the current runspace state can or should be transferred to the loop runspaces, for example defined global variables are problematic since they are likely not thread safe.

Example:
```powershell
New-PSDrive -Name ZZ -PSProvider FileSystem -Root c:\temp
1..1 | ForEach-Object -Parallel -UseCurrentState {
    # Should have access to ZZ drive
    dir ZZ:
}

Import-Module -Name c:\temp\Modules\MyModule.psd1
1..1 | ForEach-Object -Parallel -UseCurrentState {
    # Should have access to MyModule functions
    Get-MyInfo -Name Hello
}
```

This needs to be opt in since it will be a performance hit.  Some users will likely be unhappy in the performance degradation for simple scenarios.

# Proposed technical implementation details (optional)

**Comment by PaulHigin** [comment]
This is a response to other issues created against foreach -parallel, where there is a reasonable expectation that the script block is run in the current state.
#12239

**Comment by iSazonov** [comment]
Perhaps current state should be default and new switch "-UseDefaultRunspaceState".

**Comment by somescout** [comment]
May I suggest that if a -Parallel switch is specified, it is expected that only one thing changed - that For-EachObject become parallel, not that behavior of code block become different.

**Comment by PaulHigin** [comment]
@somescout  That would be ideal, but is not possible.  There are too many issues.  One is performance; replicating runspace state is time and resource intensive, and many times not even necessary.  Another is multi-threaded issues.  Much single thread code simply doesn't work in multi-threaded environments.

Like anything, there are trade offs to be made.   I feel the best solution is to provide options along with guidance to help the user know and understand the optional benefits and trade offs.

**Comment by somescout** [comment]
> One is performance; replicating runspace state is time and resource intensive, and many times not even necessary
Premature optimization. If there will be significant performance issues, it can be solved by "-Isolated" switch. But if isolation is a default - it will cause too much unpredictable behavior in parallelized code (you add -Parallel switch and your code silently produce wrong result - worst possible behavior).

> Another is multi-threaded issues. Much single thread code simply doesn't work in multi-threaded environments.
Then there should be separated parallelization commandlet, like Start-Job (13 competing standards...).

> I feel the best solution is to provide options along with guidance to help the user know and understand the optional benefits and trade offs.
Is it really worth adding a another feature, the problems with which can be found only by trial and error? (People Don't Read Instructions)

**Comment by PaulHigin** [comment]
Linking in related issue: #12313

**Comment by PaulHigin** [comment]
Per, #12378, we should also consider allowing passing in script block variables if the client runspace state is replicated.  The script block would be re-created/re-parsed within the new runspace so that it will have affinity to that runspace and run correctly.  We still have potential threading issues of any shared variable referenced in the script block, but variables and multi-threading are potential problems anyway, which a user needs to be aware of.

**Comment by iSazonov** [comment]
So we could have three options to cover all scenarios:
- clean runspace
- clean runspace with parameters
- replicated runspace with parameters

**Comment by powercode** [comment]
There will also be issues with PowerShell classes defined in the current runspace. They will not exist in the new runspaces. Hard problem, for sure.

**Comment by theJasonHelmick** [comment]
@iSazonov and @somescout - I've been thinking about the default sunspace - should by default Foreach -parallel use current (dirty) runspace or, by default, use default (clean) runspace.  It seems to me that as an admin, I would expect my current (dirty) runspace to be available when I use the -parallel switch.  While this makes intuitive sense for an admin - It just works - I wonder if the performance impact (which may be incurred for no real reason) have such a negative impact that admins would no longer see a benefit to -parallel. With this in mind, I tend to agree with @PaulHigin that the sunspace is clean.  Thoughts?

**Comment by PaulHigin** [comment]
I think I like `sunspace` better than `runspace` :).  But this is a great question.  The current runspace running ForEach-Object -Parallel could have accumulated a lot of state that is not needed for running the parallel script blocks, and transferring that state to each parallel runspace could have a major performance impact.  So my inclination is to make state transfer opt-in.  I haven't yet thought about exactly what state should be transferred.

**Comment by iSazonov** [comment]
I think `opt-in` in right direction. We could have options that transfer - from fastest clean runspace from a pool to slowest one with full features.
As side notice, we have an issue for runspace performance creation without any progress but we could investigate this too.

**Comment by somescout** [comment]
@theJasonHelmick, I, actually, don't know what to think now: it's not just runspace problem, there also issue with non-thread safe objects that can't be easily replicated to runspace:
```powershell
$obj = New-Object -ComObject Some.Object
$array | % {
  $obj.DoSomething($_)
}
```
It -Parallel is used, it will break script regardless of dirty or clean runspace used.

May be it is not a good idea in first place, to try to implement it as a part of ForEach-Object - a separate commandlet, or, may be, Start-Job with -AsRunspace and something to throttle jobs. Because everyone (who used Start-Job) is familiar how it handle local variables so there will be no broken expectations.

**Comment by vexx32** [comment]
@somescout there's already `Start-ThreadJob` with all of that, it's been available since 6.x iirc. 🙂

**Comment by PaulHigin** [comment]
@somescout  Yeah, we have been aware of multi-threaded issues with ForEach-Object -Parallel since the beginning, but ultimately felt it was still worth doing.  But script writers need to have some knowledge and use best practices, such as thread safe types if shared over parallel running script blocks.

**Comment by PaulHigin** [comment]
This work should include ensuring that experimental features are correctly enabled in the parallel running runspaces.  
#14602

**Comment by mcascone** [comment]
I vote for providing the option of runspace (sunspace) copying, perhaps not as default, so that we can use it, and take on the responsibility of dealing with the known gaps. Other issues aside, not being able to use the working runspace is worse than having to be aware of edge cases where it might be unsafe/unreliable. In most cases, especially when the user is aware of the caveats, the benefits would far outweigh the downsides. Thanks for reading.

**Comment by PaulHigin** [comment]
Also consider extending PSScriptRoot into each foreach scriptblock (Issue #15302).  Not sure yet if this makes sense for all cases.

**Comment by PaulHigin** [comment]
Also consider built-in variables (Issue #14672).

**Comment by whitTech** [comment]
> @iSazonov and @somescout - I've been thinking about the default sunspace - should by default Foreach -parallel use current (dirty) runspace or, by default, use default (clean) runspace. It seems to me that as an admin, I would expect my current (dirty) runspace to be available when I use the -parallel switch. While this makes intuitive sense for an admin - It just works - I wonder if the performance impact (which may be incurred for no real reason) have such a negative impact that admins would no longer see a benefit to -parallel. With this in mind, I tend to agree with @PaulHigin that the sunspace is clean. Thoughts?

As another admin, I also expect my runspace to be (very) dirty. And while I love performance just as much as the next admin (I mean it, please don't stop optimising for performance!), I've noticed parallelisation performance concerns around these parts often tend to be significantly exaggerated (I've personally found it to be incredibly fast, despite the abundance of caution stated [that the man himself](https://devblogs.microsoft.com/powershell/powershell-foreach-object-parallel-feature) would have you believe. @PaulHigin, you don't give yourself enough credit for your amazing work!).

These days NVMe SSD's are ubiquitous and we have more cores/threads than we know what to do with. What kind of real-world performance hits are we actually talking about in each given scenario? The hits would have to be galactic in impact to have me switching away from parallelisation when I'm applying WSUS patches to computers across my domain that can take an hour to apply (and yes, I use parallelisation for very small things too, it's equally awesome). 

With regard to the opt-in/opt-out debate, when I'm solving a problem, the priority order is as follows: Efficacy > Reliability > Performance, not the other way around. I want to build an effective solution as simply as possible, _then_ I will mess about with performance and other aspects if necessary. Therefore, assuming performance isn't genuinely crippled, I do hope end-user simplicity prevails and we opt for an opt-out approach, if possible.

**Comment by PaulHigin** [comment]
Thanks for you input and kind words.  Unfortunately, this task has been deprioritized and I have been busy working on other things.  But I still hope to get back to this in the Spring.

**Comment by whitTech** [comment]
No worries at all. 

The issue forced me to properly develop my modules and ensure they're properly distributed throughout my environments. More work, but I'm glad I have finally gotten around to it. 

Thanks again, @PaulHigin. You and the other Powershell developers deserve much more than a few kind words for the effort you contribute to this wonderful tool.

**Comment by Luiz-Monad** [comment]
What would find most useful is when using something like `$using:scriptBlock`, parallel-for-each would copy the scriptblock as a pure function in a new closure to the new state. That way using parallel would be more useful in a lot of scenarios when you just want to reuse functions, that are probably pure functions or closures with local state.
 The serialization/deserialization system ought to be able to handle that.

**Comment by noamper** [comment]
Commenting to keep this open. I would enthusiastically opt-in to a dirty runspace.
